File: top-interview-questions-medium/Trees and Graphs/binary-tree-inorder-traversal/94.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// var inorderTraversal = function(root) {
//   const result = []
//   const helper = function(root, array) {
//     if (!root) return
//     if (root.left) helper(root.left, array)
//     array.push(root.val)
//     if (root.right) helper(root.right, array)
//
//   }
//   helper(root, result)
//   return result
// }

var inorderTraversal = function(root) {
  const result = []
  const stack = []
  let cur = root
  while (cur || stack.length) {
    while (cur) {
      stack.push(cur)
      cur = cur.left
    }
    cur = stack.pop()
    result.push(cur.val)
    cur = cur.right
  }
  return result
}

File: top-interview-questions-medium/Trees and Graphs/binary-tree-inorder-traversal/94.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    # def inorderTraversal(self, root):
    #     """
    #     :type root: TreeNode
    #     :rtype: List[int]
    #     """
    #     result = []
    #     self.helper(root, result)
    #     return result
    #
    # def helper(self, root, array):
    #     if not root: return
    #     if root.left: self.helper(root.left, array)
    #     array.append(root.val)
    #     if root.right: self.helper(root.right, array)

    def inorderTraversal(self, root):
        result = []
        stack = []
        cur = root
        while cur or len(stack):
            while cur:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            result.append(cur.val)
            cur = cur.right
        return result
